packages/web-components/fast-element/src/template-compiler.ts

```diff
@@ -7,14 +7,46 @@ import { ExecutionContext, Binding } from "./observation/observable";
 type InlineDirective = Directive & {
     targetName?: string;
     binding: Binding;
-    targetAtContent();
+    targetAtContent(): void;
 };
 
-const compilationContext = { locatedDirectives: 0, targetIndex: -1 };
+class CompilationContext {
+    public targetIndex!: number;
+    public behaviorFactories!: BehaviorFactory[];
+    public directives: ReadonlyArray<Directive>;
+
+    public addFactory(factory: BehaviorFactory): void {
+        factory.targetIndex = this.targetIndex;
+        this.behaviorFactories.push(factory);
+    }
+
+    public captureContentBinding(directive: BindingDirective): void {
+        directive.targetAtContent();
+        this.addFactory(directive);
+    }
+
+    public reset(): void {
+        this.behaviorFactories = [];
+        this.targetIndex = -1;
+    }
+
+    public release(): void {
+        sharedContext = this;
+    }
+
+    public static borrow(directives: ReadonlyArray<Directive>): CompilationContext {
+        const shareable = sharedContext || new CompilationContext();
```

Comment:
So shareable can either be the `CompilationContext` or an instance of `CompilationContenxt`?